libs/MobileSync/src/com/salesforce/androidsdk/mobilesync/manager/CleanSyncGhostsTask.kt

```diff
@@ -24,69 +24,73 @@
  * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
  * POSSIBILITY OF SUCH DAMAGE.
  */
-package com.salesforce.androidsdk.mobilesync.manager;
+package com.salesforce.androidsdk.mobilesync.manager
 
-import com.salesforce.androidsdk.analytics.EventBuilderHelper;
-import com.salesforce.androidsdk.mobilesync.target.SyncDownTarget;
-import com.salesforce.androidsdk.mobilesync.util.MobileSyncLogger;
-import com.salesforce.androidsdk.mobilesync.util.SyncState;
-
-import org.json.JSONException;
-import org.json.JSONObject;
+import com.salesforce.androidsdk.analytics.EventBuilderHelper
+import com.salesforce.androidsdk.mobilesync.manager.SyncManager.CleanResyncGhostsCallback
+import com.salesforce.androidsdk.mobilesync.manager.SyncManager.SyncUpdateCallback
+import com.salesforce.androidsdk.mobilesync.target.SyncDownTarget
+import com.salesforce.androidsdk.mobilesync.util.MobileSyncLogger
+import com.salesforce.androidsdk.mobilesync.util.SyncState
+import org.json.JSONException
+import org.json.JSONObject
 
 /**
  * Runnable class responsible for running clean sync ghosts operation
  */
-public class CleanSyncGhostsTask extends SyncTask {
-
-    // Constants
-    private static final String TAG = "CleanSyncGhostsTask";
-
-    private final SyncManager.CleanResyncGhostsCallback cleanSyncCallback;
-
-    public CleanSyncGhostsTask(SyncManager syncManager, SyncState sync, SyncManager.CleanResyncGhostsCallback callback) {
-        super(syncManager, sync, null);
-        this.cleanSyncCallback = callback;
-    }
-
-    @Override
-    protected void updateSync(SyncState sync, SyncState.Status status, int progress, SyncManager.SyncUpdateCallback callback) {
+class CleanSyncGhostsTask(
+    syncManager: SyncManager,
+    sync: SyncState,
+    private val cleanSyncCallback: CleanResyncGhostsCallback?
+) : SyncTask(syncManager, sync, null) {
+    override fun updateSync(
+        sync: SyncState,
+        status: SyncState.Status,
+        progress: Int,
+        callback: SyncUpdateCallback?
+    ) {
         // Not a true sync
         // Leaving sync state alone
     }
 
-    @Override
-    protected void runSync() throws Exception {
```

Comment:
Is it significant that this is no longer open to subclass override?

Reply:
Great question about changes to API or behavior that isn't part of MSDK's documented expectation. Thinking of other libraries, like Nimbus, there has been a significant impact to internal consumers when subtle things like this change. Many consumers also use the library in ways that weren't intended, but were heavily committed to it.

The compatibility and release question wasn't fully resolved with Nimbus. From personal experience, we mitigated in several ways that worked in the long run.

One, we did on a case by case basis either restore a changed API or require a consumer to adopt it. That required a lot of post-release discussion after Nimbus 5 for Android.

Two, there was a case where we patched the previous release of Nimbus 4 with a fix required by a consumer where their release was significantly impacted. The consumer later adopted the latest version.

So, I believe changing undocumented API or behavior that is internal and not part of MSDK's acceptance criteria in 11.1 is acceptable so we can move the library forward. We might just need to be willing to patch an 11.0.2 in a worst case scenario or a quick 11.1.1 to provide support for consumers. As you've been doing, we hold changing documented, public API until 12.x. Think that works?

Sorry, long answer based on Nimbus experiences 🏁

Reply:
👍